### PR TITLE
gh-113024: C API: Add PyObject_GenericHash() function

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
* Issue: gh-113024

----
📚 Documentation preview 📚: https://cpython-previews--113025.org.readthedocs.build/

### CODE REVIEW

**Comment by vstinner** [comment]
> serhiy-storchaka requested your review on this pull request.

I replied on the issue.

File: Doc/c-api/hash.rst [hunk 1]
```diff
@@ -59,3 +59,13 @@ See also the :c:member:`PyTypeObject.tp_hash` member.
    The function cannot fail: it cannot return ``-1``.
 
    .. versionadded:: 3.13
+
+.. c:function:: Py_hash_t PyObject_GenericHash(PyObject *obj)
+
+   Generic hashing function that is meant to be put into a type
+   object's ``tp_hash`` slot.
+   It's result only depends on the object's identitity.
```
Comment (vstinner) [inline Doc/c-api/hash.rst]:
```suggestion
   Its result only depends on the object's identity.
```

Can you please add a link to built-in id() function here?

File: Doc/c-api/hash.rst [hunk 2]
```diff
@@ -59,3 +59,13 @@ See also the :c:member:`PyTypeObject.tp_hash` member.
...
+
+   In CPython it is equivalent to :c:func:`Py_HashPointer`.
```
Comment (vstinner) [inline Doc/c-api/hash.rst]:
```suggestion
   In CPython, it is equivalent to :c:func:`Py_HashPointer`.
```

You might use ``.. impl-detail::`` here.

File: Lib/test/test_capi/test_abstract.py [hunk 3]
```diff
@@ -1000,6 +1000,12 @@ def test_number_check(self):
         self.assertTrue(number_check(0.5))
         self.assertFalse(number_check("1 + 1j"))
 
+    def test_object_generichash(self):
+        # Test PyObject_GenericHash()
+        generichash = _testcapi.object_generichash
+        for obj in object(), 1, 'string', []:
+            self.assertEqual(generichash(obj), object.__hash__(obj))
```
Comment (vstinner) [inline Lib/test/test_capi/test_abstract.py:1008]: Maybe test also that it's equal to hash(id(obj))?

Comment (serhiy-storchaka) [reply Lib/test/test_capi/test_abstract.py:1008]: It is an implementation detail. It should be tested whether `object.__hash__` is tested, if tested at all.

File: Doc/c-api/typeobj.rst [hunk 4]
```diff
@@ -883,6 +883,10 @@ and :c:data:`PyType_Type` effectively act as defaults.)
    :c:member:`~PyTypeObject.tp_richcompare` and :c:member:`~PyTypeObject.tp_hash`, when the subtype's
    :c:member:`~PyTypeObject.tp_richcompare` and :c:member:`~PyTypeObject.tp_hash` are both ``NULL``.
 
+   **Default:**
+
+   :c:data:`PyBaseObject_Type` uses :c:func:`PyObject_GenericHash`.
```
Comment (vstinner) [inline Doc/c-api/typeobj.rst:888]: There is a warning: c:data reference target not found: PyBaseObject_Type

**Review by vstinner — APPROVED** [review]
LGTM, but there is a doc warning.

Also, it might be nice to add a link to the id() function in PyObject_GenericHash() documentation.

File: Doc/c-api/typeobj.rst (hunk 4 above)
Comment (serhiy-storchaka) [reply Doc/c-api/typeobj.rst:888]: It is not new. There are 14 unresolved references to PyBaseObject_Type. The preferable solution is to add a target for it, but it is not so simple, and is a different issue anyway.

**Review by serhiy-storchaka — COMMENTED** [review]
"Identity" is not the same as the result of `id()`. I answered in more detail in https://github.com/python/cpython/issues/113024#issuecomment-2003731200.

**Review by vstinner — APPROVED** [review]
LGTM

**Pull request merged**